src/Conversion/ONNXToKrnl/Tensor/Tile.cpp

```diff
@@ -0,0 +1,245 @@
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXTileOpAdaptor operandAdaptor(operands);
+    ONNXTileOp tileOp = llvm::cast<ONNXTileOp>(op);
+    auto loc = op->getLoc();
+
+    // get input operands, shapes, and rank
+    Value input = operandAdaptor.input();
+    auto inputShape = input.getType().cast<MemRefType>().getShape();
+    int64_t inputRank = inputShape.size();
+    Value repeats = operandAdaptor.repeats();
+
+    // get output info
+    auto resultOperand = tileOp.output();
+    auto outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    auto outputMemRefShape = outputMemRefType.getShape();
+    int64_t outputRank = outputMemRefShape.size();
+
+    bool insertDealloc = checkInsertDealloc(op);
+    Value alloc;
+    if (hasAllConstantDimensions(outputMemRefType))
+      alloc =
+          insertAllocAndDealloc(outputMemRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDeallocForTile(
+          outputMemRefType, loc, rewriter, insertDealloc, input, repeats);
+
+    // Define loops and iteration trip counts (equivalent to size of output)
+    std::vector<Value> originalLoops;
+    defineLoops(rewriter, loc, originalLoops, outputRank);
+    KrnlIterateOperandPack pack(rewriter, originalLoops);
+    for (int ii = 0; ii < outputRank; ++ii)
+      addDimensionToPack(rewriter, loc, pack, alloc, ii);
+
+    // Create the loops
+    auto iterateOp = rewriter.create<KrnlIterateOp>(loc, pack);
+    Block &iterationBlock = iterateOp.bodyRegion().front();
+
+    // Now perform the insertions into the body of the just generated loops.
+    // Insert instructions inside the KernelIterateOp body.
+    rewriter.setInsertionPointToStart(&iterationBlock);
+
+    // Handle the operations.
+
+    // This implementation is to iterate the output tensor.
+    // The store has simple affine subscript expression.
+    // Alternative implementation is to iterate the input tensor and repeats.
+    // The load of elements in input tensor can be reused explicitly.
+    // But the subscript of store is not contigous, or even not affine.
+    // Alternative implementation can be found at the end of this file.
+    SmallVector<Value, 4> inputMemRefVal;
+    for (int i = 0; i < outputRank; ++i) {
+      Value inputDimSizeVal;
+      if (inputShape[i] == -1)
+        inputDimSizeVal = rewriter.create<DimOp>(loc, input, i);
+      else
+        inputDimSizeVal = emitConstantOp(
+            rewriter, loc, rewriter.getIndexType(), inputShape[i]);
+      auto loopVarVal = iterationBlock.getArguments()[i];
+      auto exprVal =
+          rewriter.create<UnsignedRemIOp>(loc, loopVarVal, inputDimSizeVal);
+      inputMemRefVal.emplace_back(exprVal);
+    }
+
+    // Load the value from input
+    // Tried to use affine load when the input has constant shape
+    // But got runtime complaint, perhaps due ot RemIOp
+    auto inputVal = rewriter.create<LoadOp>(loc, input, inputMemRefVal);
```

Comment:
mod is ok for affine maps:

affine-expr ::= `(` affine-expr `)`
              | affine-expr `+` affine-expr
              | affine-expr `-` affine-expr
              | `-`? integer-literal `*` affine-expr
              | affine-expr `ceildiv` integer-literal
              | affine-expr `floordiv` integer-literal
              | affine-expr `mod` integer-literal
              | `-`affine-expr
              | bare-id
              | `-`? integer-literal

The MLIR way would be to generate a map when the TILED DIMENSIONS are constant sized.

val = affine.load(input, output indices, affine_map<(o0, o1, o2) -> (o0, o1 mod tile1, o2 mod tile2)>)

You would need a memref_cast

%4 = memref_cast %1 : memref<12x4xf32, offset:?, strides: [?, ?]> to
                      memref<12x4xf32, offset:5, strides: [4, 1]>

%4 = memref_cast %1 : memref<S0, S1, S2> to memref<S0, S1 * tile1, S2 * tile2, affine_map<d0, d1 mod tile1, d2 mod tile2>>

Could you try if this work? And if it does not, at the very least we should ask the MLIR folks why it should not work...

Reply:
The problem is caused by not using AffineExpr, not the operation.